core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/ArrayBindingSet.java

```diff
@@ -115,73 +89,79 @@ public ArrayBindingSet(ArrayBindingSet toCopy, String... names) {
-		return null;
+		return (v, a) -> {
+			a.values[index] = v;
+			a.whichBindingsHaveBeenSet[index] = true;
+			a.clearCache();
+		};
 	}
 
 	public BiConsumer<Value, ArrayBindingSet> getDirectAddBinding(String bindingName) {
-		for (int i = 0; i < this.bindingNames.length; i++) {
-			if (bindingNames[i].equals(bindingName)) {
-				final int idx = i;
-				return (v, a) -> {
-					assert !a.whichBindingsHaveBeenSet[idx] : "variable already bound: " + bindingName;
-					a.values[idx] = v;
-					a.whichBindingsHaveBeenSet[idx] = true;
-					a.clearCache();
-				};
-			}
+		int index = getIndex(bindingName);
+		if (index == -1) {
+			assert false : "variable not known to ArrayBindingSet : " + bindingName;
+			return null;
 		}
-		assert false : "variable not known to ArrayBindingSet : " + bindingName;
-		return null;
+		return (v, a) -> {
+			assert !a.whichBindingsHaveBeenSet[index] : "variable already bound: " + bindingName;
+			a.values[index] = v;
+			a.whichBindingsHaveBeenSet[index] = true;
+			a.clearCache();
+		};
+
 	}
 
-	public Function<ArrayBindingSet, Binding> getDirectGetBinding(String variableName) {
-		for (int i = 0; i < this.bindingNames.length; i++) {
-			if (bindingNames[i].equals(variableName)) {
-				final int idx = i;
-				return (a) -> {
-					Value value = a.values[idx];
-					if (value != null) {
-						return new SimpleBinding(variableName, value);
-					} else {
-						return null;
-					}
-				};
-			}
+	public Function<ArrayBindingSet, Binding> getDirectGetBinding(String bindingName) {
+
+		int index = getIndex(bindingName);
+		if (index == -1) {
+			return null;
 		}
-		return null;
+		return (a) -> {
+			Value value = a.values[index];
+			if (value != null) {
+				return new SimpleBinding(bindingName, value);
+			} else {
+				return null;
+			}
+		};
+
 	}
 
-	public Function<ArrayBindingSet, Value> getDirectGetValue(String variableName) {
-		for (int i = 0; i < this.bindingNames.length; i++) {
-			if (bindingNames[i].equals(variableName)) {
-				final int idx = i;
-				return (a) -> {
-					return a.values[idx];
-				};
-			}
+	public Function<ArrayBindingSet, Value> getDirectGetValue(String bindingName) {
+		int index = getIndex(bindingName);
+		if (index == -1) {
+			return null;
 		}
-		return null;
+		return (a) -> a.values[index];
+
 	}
 
 	public Function<ArrayBindingSet, Boolean> getDirectHasBinding(String bindingName) {
-		for (int i = 0; i < this.bindingNames.length; i++) {
+		int index = getIndex(bindingName);
+		if (index == -1) {
+			return null;
+		}
+		return (a) -> a.whichBindingsHaveBeenSet[index];
+	}
+
+	private int getIndex(String bindingName) {
+		for (int i = 0; i < bindingNames.length; i++) {
+			if (bindingNames[i] == bindingName) {
+				return i;
+			}
+		}
+		for (int i = 0; i < bindingNames.length; i++) {
 			if (bindingNames[i].equals(bindingName)) {
-				final int idx = i;
-				return (a) -> a.whichBindingsHaveBeenSet[idx];
+				return i;
 			}
 		}
-		return null;
+		return -1;
```

Comment:
getIndex first checks for reference equality, if it doesn't find a match it falls back to .equals(...) and if it finally doesn't find anything it returns -1.